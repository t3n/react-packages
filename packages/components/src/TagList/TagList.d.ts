import React from 'react';
import { TagColorVariant } from '../Tag/Tag';
export interface TagListProps {
    tags: JSX.Element[];
    collapseAfter: number;
    small?: boolean;
    colorVariant?: TagColorVariant;
    initialCollapsed?: boolean;
}
export declare const TagList: React.FC<TagListProps>;
