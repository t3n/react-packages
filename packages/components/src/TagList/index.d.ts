import TagList from './TagList';
export * from './TagList';
export default TagList;
