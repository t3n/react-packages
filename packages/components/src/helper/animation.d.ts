import { ThemeProps } from '@t3n/theme';
export declare const backgroundAnimation: import("styled-components").FlattenInterpolation<import("styled-components").ThemedStyledProps<ThemeProps, any>>;
