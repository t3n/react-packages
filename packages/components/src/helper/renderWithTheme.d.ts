import 'jest-styled-components';
import React from 'react';
import { RenderOptions } from '@testing-library/react';
import { Theme } from '@t3n/theme';
interface OptionalTheme {
    theme?: Theme;
}
export declare const renderWithTheme: (ui: React.ReactElement, { theme, ...options }: RenderOptions & OptionalTheme) => import("@testing-library/react").RenderResult<typeof import("@testing-library/dom/types/queries"), HTMLElement>;
export {};
