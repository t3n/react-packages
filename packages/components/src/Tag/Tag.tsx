import React from 'react';
import styled, { css } from 'styled-components';
import { color, margin, MarginProps, padding, variant } from 'styled-system';

import { composeTextStyle, Theme } from '@t3n/theme';

export type TagColorVariant =
  | 'primary'
  | 'secondary'
  | 'inverse'
  | 'black'
  | 'highlight'
  | 'warning'
  | 'notice'
  | 'success'
  | 'error';

export interface TagProps extends MarginProps {
  colorVariant?: TagColorVariant;
  link?: string;
  small?: boolean;
  icon?: JSX.Element;
  onClick?: () => void;
}

interface StyledTagProps {
  theme: Theme;
  variant: TagColorVariant;
  link?: string;
  small: boolean;
  clickable: boolean;
}

const StyledTag = styled.div.attrs((props: StyledTagProps) => ({
  href: props.link ? props.link : undefined,
}))`
  display: inline-flex;
  align-items: center;
  border-radius: 30px;
  ${margin};
  ${({ theme, small }: StyledTagProps) =>
    composeTextStyle({ theme, textStyle: small ? 'small' : 'regular' })}
  ${({ theme, small }: StyledTagProps) =>
    padding({ theme, px: small ? 2 : 3, py: small ? '2px' : 1 })}

  ${variant({
    variants: {
      primary: {
        bg: 'background.primary',
        color: 'text.primary',
      },
      secondary: {
        bg: 'background.secondary',
        color: 'text.primary',
      },
      inverse: {
        bg: 'background.inverse',
        color: 'text.inverse',
      },
      black: {
        bg: 'shades.black',
        color: 'text.inverse',
      },
      highlight: {
        bg: 'background.highlight',
        color: 'text.inverse',
      },
      error: {
        bg: 'background.highlight',
        color: 'text.inverse',
      },
      warning: {
        bg: 'feedback.warn',
        color: 'text.primary',
      },
      notice: {
        bg: 'feedback.notice',
        color: 'text.primary',
      },
      success: {
        bg: 'feedback.success',
        color: 'text.primary',
      },
    },
  })}

  svg {
    ${({ theme }) => margin({ theme, mt: 1, ml: 2 })}
    fill: ${({ theme, variant: variantStyle }: StyledTagProps) =>
      variantStyle === 'inverse' ||
      variantStyle === 'black' ||
      variantStyle === 'highlight' ||
      variantStyle === 'error'
        ? theme.colors.text.inverse
        : theme.colors.text.primary}
  }

  ${({ theme, clickable }: StyledTagProps) =>
    clickable &&
    css`
      text-decoration: none;
      &:hover {
        cursor: pointer;
        ${color({ theme, color: 'brand.white', bg: 'brand.red' })};

        svg {
          fill: ${theme.colors.brand.white};
        }
      }
    `}
`;

const Tag: React.FC<TagProps> = ({
  children,
  link,
  colorVariant,
  onClick,
  icon,
  small,
  ...rest
}) => (
  <StyledTag
    as={link ? 'a' : 'div'}
    link={link}
    clickable={!!link || !!onClick}
    variant={colorVariant || 'secondary'}
    onClick={onClick}
    small={small || false}
    {...rest}
  >
    {children}
    {icon}
  </StyledTag>
);

export default Tag;
