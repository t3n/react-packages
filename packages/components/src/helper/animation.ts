import { css, keyframes } from 'styled-components';

import { ThemeProps } from '@t3n/theme';

const backgroundAnimationKeyframes = ({ theme }: ThemeProps) => keyframes`
  0% {
    background-color: ${theme.colors.shades.grey204}
  }

  50% {
    background-color: ${theme.colors.shades.grey244}
  }

  100% {
    background-color: ${theme.colors.shades.grey204}
  }
`;

// eslint-disable-next-line import/prefer-default-export
export const backgroundAnimation = css`
  animation: ${backgroundAnimationKeyframes} 2s linear infinite;
`;
