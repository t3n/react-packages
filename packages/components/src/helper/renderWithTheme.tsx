/* eslint-disable import/no-extraneous-dependencies */
import 'jest-styled-components';

import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { Theme, theme as originalTheme } from '@t3n/theme';

interface OptionalTheme {
  theme?: Theme;
}

// eslint-disable-next-line import/prefer-default-export
export const renderWithTheme = (
  ui: React.ReactElement,
  { theme, ...options }: RenderOptions & OptionalTheme
) => {
  // merge over default theme

  const mergedTheme = { ...originalTheme, ...theme };

  const Wrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={mergedTheme}>{children}</ThemeProvider>
  );

  return render(ui, { wrapper: Wrapper, ...options });
};
