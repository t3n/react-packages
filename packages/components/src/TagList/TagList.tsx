import React, { useState } from 'react';
import styled from 'styled-components';
import { layout, margin } from 'styled-system';

import { MaterialArrowBack, MaterialArrowForward } from '@t3n/icons';
import { ThemeProps } from '@t3n/theme';

import Box from '../Box';
import Tag, { TagColorVariant } from '../Tag';

export interface TagListProps {
  tags: JSX.Element[];
  collapseAfter: number;
  small?: boolean;
  colorVariant?: TagColorVariant;
  initialCollapsed?: boolean;
}

const StyledTagList = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
`;

const StyledArrowButton = styled(Tag)<{ small: boolean }>`
  ${({ theme, small }) =>
    layout({
      theme,
      width: small ? '1.6rem' : '2.1rem',
      height: small ? '1.6rem' : '2.1rem',
    })};
  padding: 0;
  ${({ theme }: ThemeProps) => margin({ theme, mb: 2 })}

  &:hover {
    transform: scale(1.3);
  }

  svg {
    margin: 0 auto;
  }
`;

const TagList: React.FC<TagListProps> = ({
  initialCollapsed,
  collapseAfter,
  small = false,
  colorVariant,
  tags,
}) => {
  const [collapsed, setCollapsed] = useState<boolean>(!!initialCollapsed);

  if (tags.length <= collapseAfter) {
    return (
      <StyledTagList>
        {tags.map((tag, idx) => (
          // eslint-disable-next-line react/no-array-index-key
          <Box display="inline-block" key={idx} mb={2}>
            {tag}
          </Box>
        ))}
      </StyledTagList>
    );
  }

  if (collapsed) {
    return (
      <StyledTagList>
        {tags.slice(0, collapseAfter).map((tag, idx) => (
          // eslint-disable-next-line react/no-array-index-key
          <Box display="inline-block" key={idx} mb={2}>
            {tag}
          </Box>
        ))}
        <StyledArrowButton
          aria-label="Mehr Tags"
          small={small}
          colorVariant={colorVariant}
          onClick={() => setCollapsed(false)}
        >
          <MaterialArrowForward />
        </StyledArrowButton>
      </StyledTagList>
    );
  }

  return (
    <StyledTagList>
      {tags.map((tag, idx) => (
        // eslint-disable-next-line react/no-array-index-key
        <Box display="inline-block" key={idx} mb={2}>
          {tag}
        </Box>
      ))}{' '}
      <StyledArrowButton
        aria-label="Weniger Tags"
        small={small}
        colorVariant={colorVariant}
        onClick={() => setCollapsed(true)}
      >
        <MaterialArrowBack />
      </StyledArrowButton>
    </StyledTagList>
  );
};

TagList.defaultProps = {
  collapseAfter: 5,
  initialCollapsed: true,
  colorVariant: 'secondary',
};

export default TagList;
