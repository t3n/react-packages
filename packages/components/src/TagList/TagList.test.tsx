/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { fireEvent } from '@testing-library/dom';

import { renderWithTheme } from '../helper/renderWithTheme';
import { Tag } from '../Tag/Tag';
import { TagList } from './TagList';

test('TagList matches snapshot', () => {
  const tagLabels = ['foo', 'bar', 'baz'];

  const { container } = renderWithTheme(
    <>
      <TagList
        collapseAfter={2}
        tags={tagLabels.map((tag) => (
          <Tag>{tag}</Tag>
        ))}
      />
    </>,
    {}
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('TagList has no toggle if there are not enough items', () => {
  const tagLabels = ['foo', 'bar', 'baz'];

  const { queryByLabelText } = renderWithTheme(
    <>
      <TagList
        collapseAfter={3}
        tags={tagLabels.map((tag) => (
          <Tag>{tag}</Tag>
        ))}
      />
    </>,
    {}
  );

  expect(queryByLabelText(/mehr/i)).toBeNull();
});

test('TagList will collapse tags', () => {
  const tagLabels = ['foo', 'bar', 'baz'];

  const { queryByText, getByLabelText } = renderWithTheme(
    <>
      <TagList
        collapseAfter={2}
        tags={tagLabels.map((tag) => (
          <Tag>{tag}</Tag>
        ))}
      />
    </>,
    {}
  );

  expect(queryByText(/baz/i)).toBeNull();

  // expand items
  const toggle = getByLabelText(/mehr/i);
  fireEvent.click(toggle);

  expect(queryByText(/baz/i)).not.toBeNull();
});
