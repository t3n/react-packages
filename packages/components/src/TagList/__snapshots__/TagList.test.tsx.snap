// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagList matches snapshot 1`] = `
.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 30px;
  font-family: "adelle-sans",sans-serif;
  font-size: 1rem;
  line-height: 1.5em;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  background-color: #f4f4f4;
  color: #2a2a2a;
}

.c2 svg {
  margin-top: 4px;
  margin-left: 8px;
  fill: #2a2a2a;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 30px;
  font-family: "adelle-sans",sans-serif;
  font-size: 1rem;
  line-height: 1.5em;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  background-color: #f4f4f4;
  color: #2a2a2a;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3 svg {
  margin-top: 4px;
  margin-left: 8px;
  fill: #2a2a2a;
}

.c3:hover {
  cursor: pointer;
  color: #fff;
  background-color: #f9423a;
}

.c3:hover svg {
  fill: #fff;
}

.c1 {
  margin-bottom: 8px;
  display: inline-block;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  width: 2.1rem;
  height: 2.1rem;
  padding: 0;
  margin-bottom: 8px;
}

.c4:hover {
  -webkit-transform: scale(1.3);
  -ms-transform: scale(1.3);
  transform: scale(1.3);
}

.c4 svg {
  margin: 0 auto;
}

@media screen and (min-width:30rem) {
  .c2 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c2 {
    line-height: 1.666em;
  }
}

@media screen and (min-width:30rem) {
  .c3 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c3 {
    line-height: 1.666em;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    display="inline-block"
  >
    <div
      class="c2"
    >
      foo
    </div>
  </div>
  <div
    class="c1"
    display="inline-block"
  >
    <div
      class="c2"
    >
      bar
    </div>
  </div>
  <div
    aria-label="Mehr Tags"
    class="c3 c4"
  >
    <svg
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
      />
    </svg>
  </div>
</div>
`;
